src/wp-includes/webfonts-api/providers/class-wp-webfonts-provider.php

```diff
@@ -0,0 +1,200 @@
+ *
+ * Individual webfonts providers should extend this class and implement.
+ *
+ * @package    WordPress
+ * @subpackage WebFonts
+ * @since      5.9.0
+ */
+
+/**
+ * Abstract class for Webfonts API providers.
+ *
+ * @since 5.9.0
+ */
+abstract class WP_Webfonts_Provider {
+
+	/**
+	 * The provider's unique ID.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @var string
+	 */
+	protected $id;
+
+	/**
+	 * The provider's root URL.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @var string
+	 */
+	protected $root_url = '';
+
+	/**
+	 * Webfonts to be processed.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @var array[]
+	 */
+	protected $webfonts = array();
+
+	/**
+	 * Array of resources hints.
+	 *
+	 * Keyed by relation-type:
+	 *
+	 *      @type string $key => @type array resource hint.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @var array
+	 */
+	protected $resource_hints = array();
+
+	/**
+	 * Whether the provider fetches external resources or not.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @var bool
+	 */
+	protected $is_external = true;
+
+	/**
+	 * Get the provider's unique ID.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @return string
+	 */
+	public function get_id() {
+		return $this->id;
+	}
+
+	/**
+	 * Get the root URL for the provider.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @return string
+	 */
+	public function get_root_url() {
+		return $this->root_url;
+	}
+
+	/**
+	 * Sets this provider's webfonts property.
+	 *
+	 * The API's Controller passes this provider's webfonts
+	 * for processing here in the provider.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @param array[] $webfonts Registered webfonts.
+	 */
+	public function set_webfonts( array $webfonts ) {
+		$this->webfonts = $webfonts;
+	}
+
+	/**
+	 * Get the CSS for the font.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @return string Webfonts CSS.
+	 */
+	abstract public function get_css();
+
+	/**
+	 * Get cached styles from a remote URL.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @param string $id               An ID used to cache the styles.
+	 * @param string $url              The URL to fetch.
+	 * @param array  $args             The arguments to pass to wp_remote_get().
+	 * @param array  $additional_props Additional properties to add to the @font-face styles.
+	 * @return string The styles.
+	 */
+	public function get_cached_remote_styles( $id, $url, array $args = array(), array $additional_props = array() ) {
+		$css = get_site_transient( $id );
+
+		// Get remote response and cache the CSS if it hasn't been cached already.
+		if ( false === $css ) {
+			$css = $this->get_remote_styles( $url, $args );
+
+			// Early return if the request failed.
+			// Cache an empty string for 60 seconds to avoid bottlenecks.
+			if ( empty( $css ) ) {
+				set_site_transient( $id, '', 60 );
+				return '';
+			}
+
+			// Cache the CSS for a month.
+			set_site_transient( $id, $css, MONTH_IN_SECONDS );
+		}
+
+		// If there are additional props not included in the CSS provided by the API, add them to the final CSS.
+		foreach ( $additional_props as $prop ) {
+			$css = str_replace(
```

Comment:
Looks like `$css` is always overwritten for the last parameter of additional props, is this intentional or an append is expected instead?

Reply:
Yes, this is the expected behavior. Why? It's searching the `$css` string for the start of the font-face CSS and then adding the props just after its opening `{`. Search and replace.

Reply:
In the case where `additional_props` has more than 1 value on it. Let's say 3, even the loop goes in all 3 items because `$css` is assigned after very loop iteration only the last iteration is the actual iteration that affects the variable `$css` due the iteration 1 iteration resets the value of `$css` same for the second and the last iteration so the last iteration is the only one affecting the variable in a sense the current code behaves as: 

```php
	$last_item = $additional_props[ count(additional_props) - 1 ];
	$css = str_replace(
		'@font-face {',
		'@font-face {' . $prop . ':' . $this->params[ $prop ] . ';',
		$css
	);
return $css;
```

Reply:
Instead of searching the string each time, build the additional props string and then when done, search and insert. Micro-optimization. 🎉

Reply:
Hmm, looking at this code, it's not used in the API. And additional properties was removed for a future enhancement. I'll remove this code to simplify it.

Reply:
Removed in this commit Fixed in this commit https://github.com/WordPress/wordpress-develop/pull/1736/commits/a16c72bef59566047a41dc132e5ce09a4207ceab. cc @aristath